[ValueTracking] Analyze `Select` in `isKnownNonEqual`.

mgudim opened the pull request:
Basic way to recursively analyze `select` in `isKnownNonEqual`: `select %c, %t, %f` is non-equal to `%x` if `%t` is non-equal to `%x` and `%f` is non-equal to `%x`.

mgudim committed:
[ValueTracking] Analyze `Select` in `isKnownNonEqual`.

Basic way to recursively analyze `select` in `isKnownNonEqual`: `select %c, %t, %f` is
non-equal to `%x` if `%t` is non-equal to `%x` and `%f` is non-equal to
`%x`.

nikic commented on llvm/lib/Analysis/ValueTracking.cpp:
```diff
@@ -3112,6 +3112,23 @@ static bool isNonEqualPHIs(const PHINode *PN1, const PHINode *PN2,
   return true;
 }
 
+static bool isNonEqualSELECT(const Value *V1, const Value *V2, unsigned Depth,
```
SELECT -> Select

mgudim replied:
Done.

nikic commented on llvm/lib/Analysis/ValueTracking.cpp:
```diff
@@ -3142,6 +3159,9 @@ static bool isKnownNonEqual(const Value *V1, const Value *V2, unsigned Depth,
     };
   }
 
+  if (isNonEqualSELECT(V1, V2, Depth, Q) || isNonEqualSELECT(V2, V1, Depth, Q))
```
This is going to do duplicate work for the two selects + same condition case.

mgudim replied:
For this case, `isNonEqualSelect` will return true and second call won't happen at all. I checked this with print statements. I am compiling `opt` in`Debug` mode and this is still true.

Or did I misunderstand your statement?

nikic replied:
What I meant here is that if isNonEqualSelect returns `false`, then the second isNonEqualSelect call with swapped operands is also going to return false. Not super important though.

mgudim replied:
Oh, ok. I can do a separate PR later and refactor this and other similar calls like `if(isNonEqualMul(V1, V2, Depth, Q) || isNonEqualMul(V2, V1, Depth, Q))`.

nikic commented on llvm/test/Analysis/BasicAA/non-equal-select.ll:
```diff
@@ -0,0 +1,33 @@
+  %arrayidx2 = getelementptr inbounds [10 x i32], ptr @G, i64 0, i64 %x
+  store i32 43, ptr %arrayidx2, align 4
+  ret void
+}
```
The case of two selects with same condition is untested.

mgudim replied:
Added more tests.

nikic commented on llvm/test/Analysis/BasicAA/non-equal-select.ll:
```diff
@@ -0,0 +1,33 @@
+; RUN: opt < %s -aa-pipeline=basic-aa -passes=aa-eval -print-all-alias-modref-info -disable-output 2>&1 | FileCheck %s
+@G = global [10 x i32] zeroinitializer, align 4
+
+; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(readwrite, argmem: none, inaccessiblemem: none) uwtable
```
Drop this comment.

mgudim replied:
done

nikic commented on llvm/test/Analysis/BasicAA/non-equal-select.ll:
```diff
@@ -0,0 +1,33 @@
+@G = global [10 x i32] zeroinitializer, align 4
+
+; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(readwrite, argmem: none, inaccessiblemem: none) uwtable
+define void @select_in_gep1(i1 %c, i64 noundef %x) {
```
noundef presumably not necessary.

mgudim replied:
done

goldsteinn commented on llvm/lib/Analysis/ValueTracking.cpp:
```diff
@@ -3112,6 +3112,23 @@ static bool isNonEqualPHIs(const PHINode *PN1, const PHINode *PN2,
+                             Depth + 1, Q) &&
+             isKnownNonEqual(SI1->getFalseValue(), SI2->getFalseValue(),
+                             Depth + 1, Q);
+  }
```
You could also check cond and ~cond case

mgudim replied:
Done.

mgudim committed:
Addressed review comments

goldsteinn commented on llvm/lib/Analysis/ValueTracking.cpp:
```diff
@@ -3112,6 +3112,39 @@ static bool isNonEqualPHIs(const PHINode *PN1, const PHINode *PN2,
+      FVal1 = SI1->getFalseValue();
+      FVal2 = SI2->getFalseValue();
+    }
+    if (match(Cond1, m_Not(m_Specific(Cond2)))) {
```
I don't this is the way to do it (likewise Cond1 == Cond2).
I think something like:

```
Pred1 = Cond1->Pred;
Pred2 = Cond2->Pred;
X1 = Cond1->Op0;
Y1 = Cond1->Op1;
X2 = Cond2->Op0;
Y2 = Cond2->Op1;
if(X1 == Y2) {
   Pred2 = getSwappedPred(Pred2);
   swap(X2, Y2);
}
if(X1 == X2 && Y1 == Y2) {
   if(Pred1 == Pred2) {
       // Cond1 == Cond2 case
   }
   if(Pred1 == getInverse(Pred2) {
      // Cond1 == ~Cond2 case.
    } 
}
```

Otherwise you are fighting canonicalizations of `not(icmp)`.

mgudim replied:
Did I understand correctly: you mean we could have something like `%Cond1 = icmp sle %x, 10` and `%Cond2 = icmp sgt %x, 10`, so no explicit `xor, -1` between the two, but two are actually always opposite. Thanks, very good point!

Let me see if there is some API for that.

goldsteinn replied:
Yeah. And for the explicit `not(icmp)` I would just match that and do `Pred = getInversed(Pred)`, then the rest of the match logic can be the same.

nikic replied:
This looks like a gratuitous over-generalization. Please revert to your initial implementation.

PS: select(not(x),T,F) gets canonicalized to select(x,F,T).

goldsteinn replied:
> This looks like a gratuitous over-generalization. Please revert to your initial implementation.
> 
> PS: select(not(x),T,F) gets canonicalized to select(x,F,T) by InstCombine and does not require explicit handling.

Not for logical `and`/`or` patterns and only for single-use.

But agreed this is necessary so if you are opposed I have no issue with reverting.

mgudim replied:
Deleted the case of opposite selects.

mgudim committed:
Deleted the case of two `select`s with opposite conditions.

mgudim committed:
Moved `isNonEqualSelect` to the end of `isKnownNonEqual`.

This is necessary so that we can analyze things like
`isKnownNonEqual(X, Y)` where `X = select ... ` and `Y = shl X, 2` for
example. In other words, we first should try to do the analysis without
"decomposing" `select` first.

mgudim commented:
@nikic @goldsteinn I added a new commit moving the `isNonEqualSelect` to the end of `isKnownNonEqual`.

This is necessary so that we can analyze things like
`isKnownNonEqual(X, Y)` where `X = select ... ` and `Y = shl X, 2` for
example. In other words, we first should try to do the analysis without
"decomposing" `select`. If that doesn't succeed then we try `isNonEqualSelect`.

nikic commented on llvm/lib/Analysis/ValueTracking.cpp:
```diff
@@ -3112,6 +3112,25 @@ static bool isNonEqualPHIs(const PHINode *PN1, const PHINode *PN2,
+    return false;
+
+  if (const SelectInst *SI2 = dyn_cast<SelectInst>(V2)) {
+    const Value *Cond1 = SI1->getCondition();
+    const Value *Cond2 = SI2->getCondition();
+    if (Cond1 == Cond2)
```
```suggestion
    if (SI1->getCondition() == SI2->getCondition())
```
Unnecessary variables.

nikic commented:
> @nikic @goldsteinn I added a new commit moving the `isNonEqualSelect` to the end of `isKnownNonEqual`.
> 
> This is necessary so that we can analyze things like `isKnownNonEqual(X, Y)` where `X = select ... ` and `Y = shl X, 2` for example. In other words, we first should try to do the analysis without "decomposing" `select`. If that doesn't succeed then we try `isNonEqualSelect`.

I don't understand what you mean here. If `isNonEqualSelect()` returns false, then all the following code still gets executed, no? It shouldn't matter (for behavior) in which order these calls are executed.

mgudim commented:
> I don't understand what you mean here. If `isNonEqualSelect()` returns false, then all the following code still gets executed, no? It shouldn't matter (for behavior) in which order these calls are executed.

Right, I just thought it's better to do cheaper  checks first.

mgudim commented:
@nikic @goldsteinn ping

nikic reviewed (approved):
LGTM
